File: JS_Submissions/dynamic_programming/example2.js
/* 
Example 2: 300. Longest Increasing Subsequence

Given an integer array nums, return the length of the longest strictly increasing subsequence.

Let's say we're at index 7 and the current value is 5. We're interested in forming a strictly increasing 
subsequence where this 5 is the final value.

Because the 5 needs to be the final value, let's check the values before the current index, at indices 0 to 6.

We find that nums[5] = 4, which of course is less than the current value. We also find that dp[5] = 3. By 
definition of dp, that means there is an increasing subsequence of length 3 that ends with nums[5].

Because nums[7] > nums[5], we can take that subsequence and just append nums[7] to it, forming an increasing 
subsequence of length 4. Now, we don't need to know what the subsequence ending at nums[5] is, but we don't care
- we are only concerned about the length.

To calculate dp[7], we need to check all the indices from 0 to 6.

How did we know the value of dp[5]? Because it was a smaller subproblem. We start by solving dp[0] 
(which must be the base case 1), then dp[1] (which we can calculate from dp[0]), and so on.
*/

/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    let dp = i => {
        if (memo[i] != -1) {
            return memo[i];
        }
        
        let ans = 1; // Base case
        for (let j = 0; j < i; j++) {
            if (nums[i] > nums[j]) {
                ans = Math.max(ans, dp(j) + 1);
            }
        }
        
        memo[i] = ans;
        return memo[i];
    }
    
    let memo = new Array(nums.length).fill(-1);
    let ans = 0;
    for (let i = 0; i < nums.length; i++) {
        ans = Math.max(ans, dp(i));
    }
    
    return ans;
};


/* 
BOTTOM UP APPROACH

var lengthOfLIS = function(nums) {
    let dp = new Array(nums.length).fill(1);
    for (let i = 0; i < nums.length; i++) {
        for (let j = 0; j < i; j++) {
            if (nums[i] > nums[j]) {
                dp[i] = Math.max(dp[i], dp[j] + 1);
            }
        }
    }
    
    return Math.max(...dp);
};

*/

// Because of the nested loop, this algorithm has a time complexity of O(n^2), 
// where nn is the length of the input array.Notice that this is because the work done at each state is linear
// with n, and there are n states.The space complexity is equal to the number of states, O(n), and can't be 
// improved in bottom - up because the recurrence relation is not static.

File: Rust_Submissions/sliding_window/example2.rs
/*
Example 2: You are given a binary string s (a string containing only "0" and "1"). You may choose up to one "0" 
and flip it to a "1". What is the length of the longest substring achievable that contains only "1"?

For example, given s = "1101100111", the answer is 5. If you perform the flip at index 2, the string becomes 
1111100111.
*/

pub fn find_length(s: &str) -> i32 {
    let (mut left, mut curr, mut ans) = (0, 0, 0);
    let s_chars: Vec<char> = s.chars().collect();
    for right in 0..s_chars.len() {
        if s_chars[right] == '0' {
            curr += 1;
        }
        
        while curr > 1 {
            if s_chars[left] == '0' {
                curr -= 1;
            }
            left += 1;
        }
        
        ans = ans.max(right - left + 1);
    }
    
    ans as i32
}

/* 
this problem runs in O(n) time, where nn is the length of s, as the work done in each loop iteration is 
amortized constant. Only a few integer variables are used as well, which means this algorithm uses O(1) space.
*/